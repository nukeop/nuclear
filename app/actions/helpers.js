import uuidv4 from 'uuid/v4';
import _ from 'lodash';

export const safeAddUuid = track => {
  const clonedTrack = _.cloneDeep(track);
  if(!_.has(track, 'uuid')) {
    clonedTrack.uuid = uuidv4();
  }
  return clonedTrack;
};
