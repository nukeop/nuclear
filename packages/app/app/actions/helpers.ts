import { v4 } from 'uuid';
import _ from 'lodash';
import { createAction } from 'redux-actions';

type ActionsBasicType = {
  [k: string]: (...payload: any) => any;
}

export type ActionsType <actions extends ActionsBasicType> = {
  [k in keyof actions]: ReturnType<actions[k]>
}

export type PayloadType <actions extends ActionsType<ActionsBasicType>> = actions[keyof actions]['payload']

export const VoidAction = (actionName: string) => createAction(actionName, () => {});

export const safeAddUuid = track => {
  const clonedTrack = _.cloneDeep(track);
  if (_.isNil(track.uuid) || track.uuid.length === 0) {
    clonedTrack.uuid = v4();
  }

  return clonedTrack;
};
