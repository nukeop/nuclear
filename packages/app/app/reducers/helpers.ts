import produce from 'immer';
import { KeyCreator, LoadableMeta } from './types';

export const startingStateMeta: LoadableMeta = {
  isLoading: false,
  isReady: false,
  hasError: false
};

export const loadingStateMeta: LoadableMeta = {
  isLoading: true,
  isReady: false,
  hasError: false
};

export const successStateMeta: LoadableMeta = {
  isLoading: false,
  isReady: true,
  hasError: false
};

export const errorStateMeta: LoadableMeta = {
  isLoading: false,
  isReady: true,
  hasError: true
};

export const handleSpreadAction = <S extends {}, P = any>(state: S, { payload }: { payload: P }) => produce(state, draft => {
  for (const key of Object.keys(payload)) {
    draft[key] = payload[key];
  }
});

export const handleLoadableActionStart = <P, S>(keyCreator: KeyCreator<P>) => (state: S, { payload }: { payload: P }) =>
  produce(state, draft => {
    draft[keyCreator(payload)] = loadingStateMeta;
  });

export const handleLoadableActionSuccess = <P, S>(keyCreator: KeyCreator<P>) => (state: S, { payload }: { payload: P }) =>
  produce(state, draft => {
    draft[keyCreator(payload)] = {
      ...successStateMeta,
      data: payload
    };
  });

export const handleLoadableActionError = <P, S>(keyCreator: KeyCreator<P>, data = []) => (state: S, { payload }: { payload: P }) =>
  produce(state, draft => {
    draft[keyCreator(payload)] = {
      ...errorStateMeta,
      error: payload,
      data
    };
  });
