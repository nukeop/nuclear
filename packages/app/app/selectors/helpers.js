import _ from 'lodash';

export const createStateSelectors = (subtreeKey, keys) => {
  return _.fromPairs(_.map(keys, key => [
    key,
    state => _.get(state, `${subtreeKey}.${key}`)
  ]));
};
