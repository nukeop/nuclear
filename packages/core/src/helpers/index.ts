export { default as PlaylistHelper } from './playlist';
