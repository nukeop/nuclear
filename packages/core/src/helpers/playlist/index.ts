import { v4 } from 'uuid';
import { Playlist, PlaylistTrack, PlaylistTrackStream } from './types';

const formatPlaylistForStorage = (name: string, tracks: Array<any>, id: string = v4(), streamSource: string = null): Playlist => {
  return {
    name,
    id,
    tracks: formatTrackList(tracks, streamSource)
  };
};

const formatTrackList = (tracks, streamSource: string = null): PlaylistTrack[] => {
  const formattedTrack: PlaylistTrack[] = [];
  if (tracks) {
    tracks.forEach(track => {
      const formattedData = extractTrackData(track, streamSource);
      if (formattedData) {
        formattedTrack.push(formattedData);
      }
    });
  }

  return formattedTrack;
};

const extractTrackData = (track, streamSource: string = null): PlaylistTrack => {
  const trackStreams = track.streams || streamSource ? [{ source: streamSource, id: track.id }] : [];

  return track && (track.name || track.title) && (!track.type || track.type === 'track') ?
    {
      artist: track.artist,
      name: track.name || track.title,
      album: track.album,
      thumbnail: track.thumbnail,
      duration: track.duration,
      uuid: track.uuid,
      streams: formatTrackStreamList(trackStreams)
    } :
    null;
};

const formatTrackStreamList = (streams): PlaylistTrackStream[] => {
  const formattedStreams: PlaylistTrackStream[] = [];
  if (streams) {
    streams.forEach(stream => {
      const formattedData = extractStreamData(stream);
      if (formattedData) {
        formattedStreams.push(formattedData);
      }
    });
  }

  return formattedStreams;
};

const extractStreamData = (stream): PlaylistTrackStream => {
  return stream && stream.source && stream.id ?
    {
      source: stream.source,
      id: stream.id,
      duration: stream.duration,
      title: stream.title,
      thumbnail: stream.thumbnail
    } :
    null;
};

export default {
  formatPlaylistForStorage,
  formatTrackList,
  extractTrackData,
  formatTrackStreamList,
  extractStreamData
};
