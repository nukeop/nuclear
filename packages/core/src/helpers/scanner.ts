import {LocalTrack} from '@nuclear/scanner';
import { NuclearMeta } from '../types';

export const scannerTrackToNuclearMeta = (track: LocalTrack): NuclearMeta => {
  return {
    ...track,
    name: track.title,
    lastScanned: +Date.now()
  };
};
