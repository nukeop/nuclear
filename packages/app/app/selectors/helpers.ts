import { RootState } from '../reducers';

type StateSelectors<S extends keyof RootState, K extends keyof RootState[S]> = {
  [key in K]: (state: RootState) => RootState[S][key];
};

export const createStateSelectors = <S extends keyof RootState, K extends keyof RootState[S]>(subtreeKey: S, keys: K[]): StateSelectors<S, K> => keys.reduce((acc, key) => {
  acc[key] = (state: RootState) => state[subtreeKey][key];
  return acc;
}, {} as StateSelectors<S, K>);
