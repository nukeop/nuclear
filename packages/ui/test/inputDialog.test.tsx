import React from 'react';
import { makeSnapshotTest } from './helpers';
import { InputDialog } from '..';
import { render } from '@testing-library/react';

describe('(Snapshot) InputDialog - open', () => {
  it('should render correctly', () => {
    render(<InputDialog 
      isOpen
      header={<h4>Input header</h4>}
      acceptLabel='Accept'
      cancelLabel='Cancel'
      initialString='Initial string'
      placeholder='Input placeholder'
      onAccept={() => {}}
      onClose={() => {}}
    />);
    // we need to use document.body because the dialog is rendered outside the component fragment
    expect(document.body).toMatchSnapshot(); 
  });
});
