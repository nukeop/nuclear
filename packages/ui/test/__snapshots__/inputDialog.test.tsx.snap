// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Snapshot) InputDialog - open should render correctly 1`] = `
<body
  class="dimmable dimmed blurring"
>
  <div />
  <div
    class="ui page modals dimmer transition visible active"
    style="display: flex;"
  >
    <div
      class="ui basic modal transition visible active"
    >
      <i
        aria-hidden="true"
        class="close icon"
      />
      <div
        class="content"
      >
        <h4>
          Input header
        </h4>
        <div
          class="ui fluid inverted input"
        >
          <input
            placeholder="Input placeholder"
            type="text"
            value="Initial string"
          />
        </div>
      </div>
      <div
        class="actions"
      >
        <button
          class="ui red basic inverted button nuclear button"
        >
          Cancel
        </button>
        <button
          class="ui green button nuclear button"
        >
          Accept
        </button>
      </div>
    </div>
  </div>
</body>
`;
